### PR TITLE
Report stats for number of sigverify thread pool threads used

### PR DESCRIPTION
**Pull request by brennanwatt**
Adding histogram metrics to understand how many threads in the rayon thread pool are being activated in sigverify to make sense of performance. This will also help ensure #25833 is behaving as expected.

Performance bench data:
Before
```
test bench_sigverify_high_packets_large_batch   ... bench:  33,265,987 ns/iter (+/- 1,381,480)
test bench_sigverify_high_packets_small_batch   ... bench:  36,075,933 ns/iter (+/- 4,672,067)
test bench_sigverify_low_packets_large_batch    ... bench:   4,954,608 ns/iter (+/- 269,857)
test bench_sigverify_low_packets_small_batch    ... bench:   4,937,189 ns/iter (+/- 275,733)
test bench_sigverify_medium_packets_large_batch ... bench:   9,073,593 ns/iter (+/- 2,934,884)
test bench_sigverify_medium_packets_small_batch ... bench:   9,352,987 ns/iter (+/- 1,051,355)
test bench_sigverify_simple                     ... bench:   5,491,902 ns/iter (+/- 813,943)
```
After
```
test bench_sigverify_high_packets_large_batch   ... bench:  33,238,112 ns/iter (+/- 915,765)
test bench_sigverify_high_packets_small_batch   ... bench:  36,729,933 ns/iter (+/- 12,555,219)
test bench_sigverify_low_packets_large_batch    ... bench:   4,947,410 ns/iter (+/- 281,600)
test bench_sigverify_low_packets_small_batch    ... bench:   4,920,535 ns/iter (+/- 178,547)
test bench_sigverify_medium_packets_large_batch ... bench:   8,541,525 ns/iter (+/- 1,071,851)
test bench_sigverify_medium_packets_small_batch ... bench:   9,651,006 ns/iter (+/- 1,203,085)
test bench_sigverify_simple                     ... bench:   5,557,156 ns/iter (+/- 1,161,243)
```

### CODE REVIEW

File: perf/src/sigverify.rs [hunk 1]
```diff
@@ -608,12 +609,18 @@ pub fn shrink_batches(batches: &mut Vec<PacketBatch>) {
     batches.truncate(last_valid_batch);
 }
 
-pub fn ed25519_verify_cpu(batches: &mut [PacketBatch], reject_non_vote: bool, packet_count: usize) {
+pub fn ed25519_verify_cpu(
+    batches: &mut [PacketBatch],
+    reject_non_vote: bool,
+    packet_count: usize,
+    active_threads_hist: Option<&mut histogram::Histogram>,
```
Comment (behzadnouri) [inline perf/src/sigverify.rs:616]: Why is this an `Option`?

File: perf/src/sigverify.rs [hunk 2]
```diff
@@ -622,33 +629,47 @@ pub fn ed25519_verify_cpu(batches: &mut [PacketBatch], reject_non_vote: bool, pa
...
+        let packet_verification_closure = |packet: &mut Packet| {
+            if !packet.meta.discard() && !verify_packet(packet, reject_non_vote) {
+                packet.meta.set_discard(true);
+            }
+            if let Some(thread_mask) = CheckedShl::checked_shl(
+                &1,
+                rayon::current_thread_index().unwrap().try_into().unwrap(),
+            ) {
+                thread_in_use_multihot.fetch_or(thread_mask, Ordering::Relaxed);
+            }
```
Comment (behzadnouri) [inline perf/src/sigverify.rs:643]: wouldn't this add extra overhead to packet verification?
Also wouldn't this break if there are more than 64 threads?

**Review by behzadnouri — COMMENTED** [review]
You can write a test locally which generates some packets and then add some annotations or debug prints to track how many threads were actually used. But I am not sure this is a good use case for metrics; why do we need to gather metrics all the time from the validators how many threads are they using? it seems like a static problem which can be measured locally.

**Comment by brennanwatt** [comment]
> You can write a test locally which generates some packets and then add some annotations or debug prints to track how many threads were actually used. But I am not sure this is a good use case for metrics; why do we need to gather metrics all the time from the validators how many threads are they using? it seems like a static problem which can be measured locally.

Definitely, I had to test locally extensively during the #25883 dev. I see two main benefits to tracking this at cluster level:
1. Validators can behave differently given the algorithm depends on number/availability of cores at time of verification. Having an idea of how much compute is being allocated towards verification could be useful. That said, we can infer most of this using packet count histogram (I say most because Rayon will not deterministically allocate threads based strictly on packet count).
2. We've discussed making the dedup and verify batch thresholds (`MAX_DEDUP_BATCH` & `MAX_SIGVERIFY_BATCH`) dynamic based on what the validator is actually capable of. However, in order to understand this, we need to understand how much parallelism is actually being deployed. I've slapped together a PoC for this at #25970 to show what I'm talking about

It seems like we can live without 1. 2 is the more compelling argument in my mind. Let me know what you think

File: perf/src/sigverify.rs (hunk 1 above)
Comment (brennanwatt) [reply perf/src/sigverify.rs:616]: There are some cases where we don't care to spin up a Stats structure and collect this information. For example, unit testing, benches, verifying votes int he vote listener (at least currently)

**Comment by behzadnouri** [comment]
> Having an idea of how much compute is being allocated towards verification could be useful.
> However, in order to understand this, we need to understand how much parallelism is actually being deployed.

We already have metrics for `cpu-stats`, `memory-stats`, etc.
You can spin up a gce machine of similar specs to run performance tests.

> We've discussed making the dedup and verify batch thresholds (MAX_DEDUP_BATCH & MAX_SIGVERIFY_BATCH) dynamic based on what the validator is actually capable of.

I am not sure this is a good direction to go.
We spin up a thread pool of the size based on the number of machine cores. After that this is a task scheduling problem which _ideally should be encapsulated in the thread-pool implementation_. The rayon documentation itself for example says that:

> Parallel iterators take care of deciding how to divide your data into tasks; it will **dynamically** adapt for maximum performance.
> https://github.com/rayon-rs/rayon/blob/99b1ff56f/README.md#parallel-iterators-and-more

Adding too many knobs and several layers of dynamic adjustment just makes things too complicated to understand and debug.

If you are seeing evidence that rayon dynamic scheduling is not doing a good enough job and cannot be fine tuned with knobs rayon already provides (`with_min_len`, etc), then perhaps would be better to raise this to rayon developers and see if either they have any suggestion or if their implementation can be improved.

Adding one layer of dynamic adjustment on top of another does not seem to me a good option.

File: perf/src/sigverify.rs (hunk 2 above)
Comment (behzadnouri) [reply perf/src/sigverify.rs:643]: > I don't believe we have any cases where we are running with that many threads, 

You can check `cpu-stats` metrics

**Comment by brennanwatt** [comment]
> > Having an idea of how much compute is being allocated towards verification could be useful.
> > However, in order to understand this, we need to understand how much parallelism is actually being deployed.
> 
> We already have metrics for `cpu-stats`, `memory-stats`, etc. You can spin up a gce machine of similar specs to run performance tests.
>
Yeah, these work well for understanding HW resource utilization/bottlenecks at the global level. I can imagine a specific case of seeking to understand Sigverify throughput (maybe verifications per second seems low) and we suspect #25883 introduced some problem :) . But again, this is not the hill I'll choose to die on because we can mostly infer this from packet count statistics.

> > We've discussed making the dedup and verify batch thresholds (MAX_DEDUP_BATCH & MAX_SIGVERIFY_BATCH) dynamic based on what the validator is actually capable of.
> 
> I am not sure this is a good direction to go. We spin up a thread pool of the size based on the number of machine cores. After that this is a task scheduling problem which _ideally should be encapsulated in the thread-pool implementation_. The rayon documentation itself for example says that:
> 
> > Parallel iterators take care of deciding how to divide your data into tasks; it will **dynamically** adapt for maximum performance.
> > https://github.com/rayon-rs/rayon/blob/99b1ff56f/README.md#parallel-iterators-and-more
> 
> Adding too many knobs and several layers of dynamic adjustment just makes things too complicated to understand and debug.
> 
> If you are seeing evidence that rayon dynamic scheduling is not doing a good enough job and cannot be fine tuned with knobs rayon already provides (`with_min_len`, etc), then perhaps would be better to raise this to rayon developers and see if either they have any suggestion or if their implementation can be improved.
> 
> Adding one layer of dynamic adjustment on top of another does not seem to me a good option.

Right, I don't disagree with what you're saying, but I'm not sure how it applies here. It seems like you're talking about how we allocate work to the various threads, but I'm referring to our load shedding thresholds. I view these as related but separate. Both policies exist today, so it comes down to our appetite to inject complexity into the load shedding algo in order to drop fewer packets on high performing machines.

I believe this was @sakridge 's idea, so I'll let him weigh in on how important this is and if we're okay with keeping static thresholds.

File: perf/src/sigverify.rs (hunk 2 above)
Comment (brennanwatt) [reply perf/src/sigverify.rs:643]: Highest core count is 96 on testnet, which I believe would mean max of 48 threads for Sigverify. Either way, future proofing is good

**Comment by brennanwatt** [comment]
Going to abandon these metrics changes. We will still pursue dynamic verification packet threshold, but we will use a strategy where we only recompute verification throughput when packet count is large and use this as a proxy for max concurrency (i.e. don't actually measure threads in use).

**Pull request closed without merging**